### PR TITLE
Adding uint4 dtype implementation

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #13

Summary:
This PR added some preliminary support for uint4 through tensor subclass and we'll continue to iterate on this

we plan to move the uint4 tensor subclass to core after it is more mature

Test Plan:
python test/dtypes/test_int4.py

Reviewers:

Subscribers:

Tasks:

Tags:

### CODE REVIEW

File: torchao/dtypes/int4.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+    @classmethod
+    def from_unpacked(cls, unpacked):
+        return UInt4Tensor(pack_uint4(unpacked))
+
+    def tolist(self):
+        return self.to(torch.uint8).tolist()
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs=None):
```
Comment (bdhirsh) [inline torchao/dtypes/int4.py:69]: One thing you're missing is a `__tensor_flatten__` and `__tensor_unflatten__` method. Without this, dynamo will graph break on all usages of your subclass. You can see an example of it here: https://github.com/pytorch/pytorch/blob/main/torch/testing/_internal/two_tensor.py

File: torchao/dtypes/int4.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+# https://github.com/drisspg/transformer_nuggets/blob/9ad3a7fc552a954eb702ade0e276b8d8e09c3db6/transformer_nuggets/quant/qlora.py#L233
+def unpack_uint4(quantized_data) -> torch.Tensor:
+    """Get the original weight from the normalized float weight format"""
+    # since we are using uint8 we will decode 2 entries per byte
+    # Shift elements down 4 and select out the bottom 4 bits
+    first_elements = (quantized_data >> 4).to(torch.uint8)
+    second_elements = (quantized_data & 0b1111).to(torch.uint8)
+    return torch.stack([first_elements, second_elements], dim=-1)
+
+def pack_uint4(uint8_data) -> torch.Tensor:
```
Comment (cpuhrsch) [inline torchao/dtypes/int4.py:45]: Any way we can just hook this up to `.to_dtype` native function or such?

File: torchao/dtypes/int4.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+    shape = uint8_data.shape
+    uint8_data = uint8_data.contiguous().view(-1)
+    return (uint8_data[::2] << 4 | uint8_data[1::2]).view(down_size(shape))
+
+class UInt4Tensor(torch.Tensor):
```
Comment (cpuhrsch) [inline torchao/dtypes/int4.py:50]: We're creating a whole new Tensor here, but really we want to extend a QTensor with a new lower precision backend? Or do we want to combine UInt4Tensor with a QTensor, where QTensor has a UInt4Tensor for storing int4 weights (next to scales etc.)?

File: torchao/dtypes/int4.py (hunk 1 above)
Comment (jerryzh168) [reply torchao/dtypes/int4.py:69]: OK fixed

File: torchao/dtypes/int4.py (hunk 2 above)
Comment (jerryzh168) [reply torchao/dtypes/int4.py:45]: yeah I hope to connect these with tensor factory functions, but not sure how to do this right now. last I heard from Ed that he is saying it's non trivial to do, I'll follow up a bit later

File: torchao/dtypes/int4.py (hunk 3 above)
Comment (jerryzh168) [reply torchao/dtypes/int4.py:50]: by QTensor do you mean https://github.com/pytorch/pytorch/blob/main/aten/src/ATen/quantized/QTensorImpl.h? I think we are moving away from this and just rely on PyTorch native dtypes now

**Comment by jerryzh168** [comment]
> The next major issue I think you'll hit is that with `torch.compile`, we don't have a recipe today for "send subclass directly to the compiler, instead of desugaring it". Most of the pieces for this to work should already exist, but let me know and once the other errors are fixed we can try to sit down together and get this to work.

oh the suggestion from Alban is that we'll desugar it for now, since triton don't have int4 support anyways and we are using handwritten custom kernels, the int4 is just for representation in frontend

File: torchao/dtypes/int4.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            self, size, stride, storage_offset = args
+            size = up_size(size)
+            new_stride = []
+            for s in stride:
+                if s != 1:
+                    # since two int4 equals to 1 byte
+                    new_stride.append(s // 2)
+                else:
+                    # wondering if we need to have 1/2 stride?
+                    new_stride.append(s)
```
Comment (jerryzh168) [inline torchao/dtypes/int4.py:140]: @bdhirsh not sure if this is correct, Ed also mentioned that we can create fractional stride (1/8) before

Comment (jerryzh168) [reply torchao/dtypes/int4.py:140]: discussed with @albanD and he says these are all talking about elements, not bytes, so updated the code and looks like it works now

Comment (albanD) [reply torchao/dtypes/int4.py:144]: Do you have the constraint that the last dimension of any int4 Tensor has to be a multiple of 2?
and the end user has to do padding themselves?

Comment (jerryzh168) [reply torchao/dtypes/int4.py:144]: yes, that's our assumption, but I haven't added explicit error checks yet

File: test/dtypes/test_int4.py [hunk 5]
```diff
@@ -0,0 +1,321 @@
...
+        for x_shape in [[2, 4], [5, 5, 5, 4], [1, 4, 4]]:
+            x = torch.randn(*x_shape)
+            m = nn.Sequential(nn.Linear(4, 16))
+            y_ref = m(x)
+            _apply_weight_only_int4_quant(m)
+            y_wo = m(x)
+            # sqnr = compute_error(y_ref, y_wo)
+            opt = torch.compile(m, mode="max-autotune")
+            # make sure it runs
+            opt(x)
```
Comment (bdhirsh) [inline test/dtypes/test_int4.py:167]: I'm seeing an assertion failure [here](https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/meta_utils.py#L474)

When I look into it more:

(1) `symbolic_context` is a `StatelessSymbolicContext`, not a `SubclassSymbolicContext`

(2) that's because when we fakeify a tensor that is a view, we unconditionally create a `StatelessSymbolicContext` for its base: https://github.com/pytorch/pytorch/blob/main/torch/_subclasses/meta_utils.py#L330

(3) However in this case, the ._base of the tensor we're fakifying is another subclass, so it should get a `SubclassSymbolicContext`

I'm going to make a min repro and file an issue for it. Joel has also been doing some work around fakifying subclasses that are views, so hopefully his work will also fix this bug.

**Comment by iseeyuan** [comment]
Are you planning to add int4 dtype from ATen C++ side?

**Comment by jerryzh168** [comment]
> Are you planning to add int4 dtype from ATen C++ side?

no we don't, we'll use bits8 instead, and encode int4 related information in ops that is using these tensors, e.g. `dequantize_int4(bits8_tensor, scale, zero_point, qmin, qmax, ...)`

File: test/dtypes/test_int4.py (hunk 5 above)
Comment (bdhirsh) [reply test/dtypes/test_int4.py:167]: @jerryzh168 this patch should also avoid the issue for you:
```
--- a/test/dtypes/test_int4.py
+++ b/test/dtypes/test_int4.py
@@ -91,7 +91,7 @@ class _WeightOnlyInt4QuantLinear(torch.nn.Linear):
             fake_in_features,
             fake_out_features,
             bias=mod.bias is not None,
-            w_int4=w_int4.t().contiguous(),
+            w_int4=torch.ops.aten.transpose_copy.default(w_int4),
             scales=scales,
         )
         new_mod.in_features = mod.in_features
```

It looks like the way that `UInt4Tensor.__torch_dispatch__` implements transpose, it always returns a contiguous tensor (is that what you want?). This means that:

(1) `w_int4.t()` will report as being contiguous (`w_int4.t().is_contiguous() == True`)

(2) the `.contiguous()` call afterwards will be a no-op

(3) the autograd engine will think that `w_int4.t().contiguous()` is a view, and not a fresh tensor. This means that the final weight in your nn module, `w_int4`, will have a `._base` field, causing it to go down a more interesting path in the fakeification logic.

One easy workaround for now is that since you know that your subclass implements transpose to return a contiguous tensor, you can just call `transpose_copy` from the user code, so autograd will know that the output does not alias the original tensor.

Another option though is that you might want to implement `transpose` with the proper semantics, so that it returns a tensor that is not contiguous.

File: test/dtypes/test_int4.py [hunk 6]
```diff
@@ -0,0 +1,321 @@
...
+
+
+class TestInt4(QuantizationTestCase):
+    def test_basic_tensor_ops(self):
+        x = UInt4Tensor(torch.tensor([
+            [0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF],
+            [0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF],
+            [0x01, 0x23, 0x45, 0x67, 0x89, 0xAB, 0xCD, 0xEF],
+        ], dtype=torch.uint8))
+        self.assertTrue(x.shape, (3, 8))
```
[the remark this answers is not in the document]
Comment (jerryzh168) [reply test/dtypes/test_int4.py]: oh right, thanks. fixed in the new commit

File: torchao/dtypes/int4.py (hunk 3 above)
Comment (jerryzh168) [reply torchao/dtypes/int4.py:83]: @cpuhrsch I guess by QTensor you mean these tensors: https://github.com/pytorch-labs/ao/blob/main/torchao/quantization/subclass.py#L178, This UInt4Tensor can compose with that (it's similar to uint8 tensor), see the example below (PerChannelSymmetricWeightUInt4Tensor)

File: torchao/dtypes/int4.py [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+    scale: float,
+    zero_point: int,
+) -> torch.Tensor:
+    input = unpack_uint4(input)
+    return (input.view(torch.uint8).to(torch.float32) - zero_point) * scale
+
+class UInt4Tensor(torch.Tensor):
+    @staticmethod
+    def __new__(cls, elem, **kwargs):
+        # TODO: uint64 here is wrong, need a real dtype.  Don't try to(int64)
```
Comment (airMeng) [inline torchao/dtypes/int4.py:86]: Hi Pytorch Team, can we introduce ```group-wise``` concept as llama.cpp and OCP MXformat do? Sub-byte tensors are always organized as groups, each group share one scale(might + zero points), or even exponent (OCP MXformat)

For details you can refer to our [proposal](https://github.com/intel/neural-speed/blob/rfcs/pytorch_lowerprecision.md)

Please let me know if helpful.

Comment (jerryzh168) [reply torchao/dtypes/int4.py:86]: yes, this PR is paving the way for these new ways of quantization etc. I'll probably just support simple per tensor int4 quantization as a starting point in this PR, but we will follow up with more changes and add groupwise/grouped quantization

Comment (airMeng) [reply torchao/dtypes/int4.py:86]: BTW, most sub-byte tensors introduce block-wise storage for accuracy/alignments/performance.
Have you considered whether it is possible to introduce a brand-new storage for sub-byte tensors?

Comment (jerryzh168) [reply torchao/dtypes/int4.py:82]: we don't need new storage I think? we can still simulate this in the tensor subclass right

File: torchao/dtypes/uint4.py [hunk 8]
```diff
@@ -0,0 +1,304 @@
...
+        scales = flattened["scales"]
+        return PerChannelSymmetricWeightUInt4Tensor(elem, scales)
+
+    @classmethod
+    def from_unpacked(cls, unpacked, scales):
+        return cls(pack_uint4(unpacked), scales)
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs=None):
+        # didin't hit this, there is a mysterious error if we try to go through this path
```
Comment (bdhirsh) [inline torchao/dtypes/uint4.py:275]: is this comment still relevant?

File: torchao/dtypes/uint4.py [hunk 9]
```diff
@@ -0,0 +1,304 @@
...
+qtensor_lib.define("quantize_per_tensor_uint4(Tensor input, float scale, int zero_point) -> Tensor")
+
+@impl(qtensor_lib, "quantize_per_tensor_uint4", "CompositeExplicitAutograd")
+def quantize_per_tensor_uint4(
+    input: torch.Tensor,
+    scale: float,
+    zero_point: int,
+) -> torch.Tensor:
+    inv_scale = 1.0 / scale
+    return pack_uint4(torch.clamp(torch.round(input * inv_scale) + zero_point, 0, 15).to(torch.uint8))
```
Comment (bdhirsh) [inline torchao/dtypes/uint4.py:65]: for inductor and export, how do you want these routines to be optimized? Is there a custom kernel for the quantize/dequantizing? Or would you prefer to for inductor to "see" this decomposition, so it can try to fuse it into a triton kernel.

File: torchao/dtypes/uint4.py [hunk 10]
```diff
@@ -0,0 +1,304 @@
...
+
+    def __hash__(self):
+        return hash(self.elem)
+
+    def __eq__(self, other):
+        return torch.equal(self.elem, other.elem)
+
+    # def __getattribute__(self, name):
+    #     if name == "dtype":
+    #         return torch.uint4
```
Comment (bdhirsh) [inline torchao/dtypes/uint4.py:112]: is uint4 a native dtype now? Does that mean we don't need any sort of custom logic here?

File: torchao/dtypes/uint4.py [hunk 11]
```diff
@@ -0,0 +1,304 @@
...
+    #     return super().__getattribute__(name)
+
+    @classmethod
+    def __torch_dispatch__(cls, func, types, args, kwargs=None):
+        if func is torch.ops.aten.view.default:
+            self, size = args
+            size = utils.infer_size(size, self.numel())
+            assert not kwargs
+            # WARNING: views not preserved
```
Comment (bdhirsh) [inline torchao/dtypes/uint4.py:121]: why aren't we preserving the view?

File: torchao/dtypes/uint4.py (hunk 8 above)
Comment (jerryzh168) [reply torchao/dtypes/uint4.py:275]: oh we can remove this now

File: torchao/dtypes/uint4.py (hunk 9 above)
Comment (jerryzh168) [reply torchao/dtypes/uint4.py:65]: I think it will be the latter for now, but maybe in the future we can have custom kernels

File: torchao/dtypes/uint4.py (hunk 10 above)
Comment (jerryzh168) [reply torchao/dtypes/uint4.py:112]: yes, that's right

File: torchao/dtypes/uint4.py (hunk 11 above)
Comment (jerryzh168) [reply torchao/dtypes/uint4.py:121]: this is copied from Edward's original PR, not exactly sure why, feels like this is creating a new tensor so we don't have view relationship? is view supposed to share storage?

File: torchao/dtypes/uint4.py [hunk 12]
```diff
@@ -0,0 +1,295 @@
...
+    zero_point: int,
+) -> torch.Tensor:
+    input = unpack_uint4(input)
+    return (input.view(torch.uint8).to(torch.float32) - zero_point) * scale
+
+class UInt4Tensor(torch.Tensor):
+    @staticmethod
+    def __new__(cls, elem, **kwargs):
+        assert elem.dtype is torch.uint8
+        assert not kwargs.get("requires_grad", False)
```
Comment (shuokay) [inline torchao/dtypes/uint4.py:81]: When will UInt4Tensor receive `requires_grad` since `__init__` only support one position argument?

Comment (jerryzh168) [reply torchao/dtypes/uint4.py:81]: oh makes sense, thanks. I'll add **kwargs to `__init__` as well

**Review by HDCharles — APPROVED** [review]
lgtm
